File: source/gameFunc/Collisions.cs
using static Defines;
using static Levels;

static class Collisions
{
    public static bool objectCollision(ObjectInfo obj1, ObjectInfo obj2)
    {
        // Get edges of both rectangles
        int left1 = (int)obj1.x;
        int up1 = (int)obj1.y;
        int right1 = left1 + obj1.objClass.width;
        int down1 = up1 + obj1.objClass.height;

        int left2 = (int)obj2.x;
        int up2 = (int)obj2.y;
        int right2 = left2 + obj2.objClass.width;
        int down2 = up2 + obj2.objClass.height;

        if (left1 > right2 || right1 < left2 || up1 > down2 || down1 < up2) return false;
        return true;
    }

    public static bool objectCollisionTop(ObjectInfo obj1, ObjectInfo obj2)
    {
        int down1 = (int)obj1.y + obj1.objClass.height;
        int up2 = (int)obj2.y;
        // Return true if collision point is within top quarter of object
        return objectCollision(obj1, obj2) && down1 - up2 < obj2.objClass.height / 4;
    }

    public enum LineDir { Horizontal, Vertical }

    public static bool lineCollision((int x, int y) start, LineDir direction, int length)
    {
        (int x, int y) vector;
        if (direction == LineDir.Horizontal) vector = (1, 0);
        else vector = (0, 1);

        for (int i = 0; i < length; i++)
        {
            if (currentLevel.getCollisionPix(MISCSCREEN, 3, start.x + vector.x * i, start.y + vector.y * i)) return true;
            if (currentLevel.getCollisionPix(MISCSCREEN, 3, start.x - vector.x * i, start.y - vector.y * i)) return true;
        }

        return false;
    }

    public static bool upCollision(ObjectInfo obj)
    {
        if (!obj.alive) return false;
        return lineCollision(((int)obj.cx, (int)obj.y),
            LineDir.Horizontal, (obj.objClass.width / 2));
    }

    public static bool downCollision(ObjectInfo obj)
    {
        if (!obj.alive) return false;
        return lineCollision(((int)obj.cx, (int)obj.y + obj.objClass.height),
            LineDir.Horizontal, (obj.objClass.width / 2));
    }

    public static bool touchingGround(ObjectInfo obj)
    {
        if (!obj.alive) return false;
        return lineCollision(((int)obj.cx, (int)obj.y + obj.objClass.height + 1),
            LineDir.Horizontal, (obj.objClass.width / 2));
    }

    public static bool leftCollision(ObjectInfo obj)
    {
        if (!obj.alive) return false;
        return lineCollision(((int)obj.x, (int)obj.cy),
            LineDir.Vertical, (obj.objClass.height / 2) - 6);
    }

    public static bool rightCollision(ObjectInfo obj)
    {
        if (!obj.alive) return false;
        return lineCollision(((int)obj.x + obj.objClass.width, (int)obj.cy),
            LineDir.Vertical, (obj.objClass.height / 2) - 6);
    }
}

File: source_original/gameFunc/collisions.c
#include "collisions.h"
#include "defines.h"
#include "objects.h"
#include "level.h"

bool objectCollision(int object1,int object2){
//grab sprite demensions (center positions and length and width)
  int w1 = object[object1].class->width;
  int h1 = object[object1].class->height;
  int x1 = object[object1].cx>>8;
  int y1 = object[object1].cy>>8;

  int w2 = object[object2].class->width;
  int h2 = object[object2].class->height;
  int x2 = object[object2].cx>>8;
  int y2 = object[object2].cy>>8;
  
  if(object1 != object2) 
      return (((x2 >= x1 - ((w1 + w2)>>1)) && (x2 <= x1 + ((w1 + w2)>>1)) && (y2 >= y1 - ((h1 + h2)>>1)) && (y2 <= y1 + ((h1 + h2)>>1))));
  return 0;
}

bool objectCollisionTop(int object1,int object2){
//grab sprite demensions (center positions and length and width)
  int w1 = object[object1].class->width;
  int h1 = object[object1].class->height;
  int x1 = object[object1].cx>>8;
  int y1 = object[object1].cy>>8;

  int w2 = object[object2].class->width;
  int h2 = object[object2].class->height;
  int x2 = object[object2].cx>>8;
  int y2 = object[object2].cy>>8;

  if(object1 != object2 && (y1+(h1>>1))-h1 < y2 - (h2))
      return (((x2 >= x1 - ((w1 + w2)>>1)) && (x2 <= x1 + ((w1 + w2)>>1)) && (y2 >= y1 - ((h1 + h2)>>1)) && (y2 <= y1 + ((h1 + h2)>>1))));

  return 0;
}


bool upCollision(int objectNum)
{
    if(!object[objectNum].alive) return 0;

    s8 playerheight=object[objectNum].class->height;
    s8 playerwidth=object[objectNum].class->width;
    int playerx = object[objectNum].cx>>8;
    int playery = (object[objectNum].y>>8) + 62;

    if(getCollisionPix(MISCSCREEN,3,playerx,playery-playerheight))return 1;
    if(getCollisionPix(MISCSCREEN,3,(playerx-(playerwidth>>1))+COLLISION_BORDER,playery-playerheight))return 1;
    if(getCollisionPix(MISCSCREEN,3,(playerx+(playerwidth>>1))-COLLISION_BORDER,playery-playerheight))return 1;
    return 0;


}

bool downCollision(int objectNum)
{
    if(!object[objectNum].alive) return 0;

    s8 playerheight=object[objectNum].class->height;
    s8 playerwidth=object[objectNum].class->width;
    int playerx = object[objectNum].cx>>8;
    int playery = (object[objectNum].y>>8) + 62;

    
    if(getCollisionPix(MISCSCREEN,3,playerx,playery))return 1;
    if(getCollisionPix(MISCSCREEN,3,(playerx-(playerwidth>>1))+COLLISION_BORDER,playery))return 1;
    if(getCollisionPix(MISCSCREEN,3,(playerx+(playerwidth>>1))-COLLISION_BORDER,playery))return 1;
    return 0;
}

bool touchingGround(int objectNum)
{
    if(!object[objectNum].alive) return 0;

    s8 playerheight=object[objectNum].class->height;
    s8 playerwidth=object[objectNum].class->width;
    int playerx = object[objectNum].cx>>8;
    int playery = (object[objectNum].y>>8) + 63;

    if(getCollisionPix(MISCSCREEN,3,playerx,playery))return 1;
    if(getCollisionPix(MISCSCREEN,3,(playerx-(playerwidth>>1))+COLLISION_BORDER,playery))return 1;
    if(getCollisionPix(MISCSCREEN,3,(playerx+(playerwidth>>1))-COLLISION_BORDER,playery))return 1;
    return 0;
}

bool leftCollision(int objectNum)
{
    if(!object[objectNum].alive) return 0;

    s8 playerheight=object[objectNum].class->height;
    s8 playerwidth=object[objectNum].class->width;
    int playerx = object[objectNum].cx>>8;
    int playery = (object[objectNum].y>>8) + 62;

    if(getCollisionPix(MISCSCREEN,3,playerx-(playerwidth>>1),playery-COLLISION_BORDER))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx-(playerwidth>>1),playery-(playerheight>>1)))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx-(playerwidth>>1),playery-(playerheight>>2)))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx-(playerwidth>>1),(playery-playerheight)+COLLISION_BORDER))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx-(playerwidth>>1),(playery-playerheight)+(playerheight>>2)+(playerheight>>3)))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx-(playerwidth>>1),(playery-playerheight)+(playerheight>>3)))return 1;
    return 0;
}

bool rightCollision(int objectNum)
{
    if(!object[objectNum].alive) return 0;

    s8 playerheight=object[objectNum].class->height;
    s8 playerwidth=object[objectNum].class->width;
    int playerx = object[objectNum].cx>>8;
    int playery = (object[objectNum].y>>8) + 62;


    if(getCollisionPix(MISCSCREEN,3,playerx+(playerwidth>>1),playery-COLLISION_BORDER))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx+(playerwidth>>1),playery-(playerheight>>1)))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx+(playerwidth>>1),playery-(playerheight>>2)))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx+(playerwidth>>1),(playery-playerheight)+COLLISION_BORDER))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx+(playerwidth>>1),(playery-playerheight)+(playerheight>>2)+(playerheight>>3)))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx+(playerwidth>>1),(playery-playerheight)+(playerheight>>3)))return 1;
    return 0;
}

bool leftCollisionLarge(int objectNum)
{
    if(!object[objectNum].alive) return 0;

    s8 playerheight=object[objectNum].class->height;
    s8 playerwidth=object[objectNum].class->width;
    int playerx = object[objectNum].cx>>8;
    int playery = (object[objectNum].y>>8) + 24;

    if(getCollisionPix(MISCSCREEN,3,playerx-(playerwidth>>1),playery-COLLISION_BORDER))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx-(playerwidth>>1),playery-(playerheight>>1)))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx-(playerwidth>>1),(playery-playerheight)+COLLISION_BORDER))return 1;
    return 0;
}

bool rightCollisionLarge(int objectNum)
{
    if(!object[objectNum].alive) return 0;

    s8 playerheight=object[objectNum].class->height;
    s8 playerwidth=object[objectNum].class->width;
    int playerx = object[objectNum].cx>>8;
    int playery = (object[objectNum].y>>8) + 32;


    if(getCollisionPix(MISCSCREEN,3,playerx+(playerwidth>>1),playery-COLLISION_BORDER))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx+(playerwidth>>1),playery-(playerheight>>1)))return 1;
    if(getCollisionPix(MISCSCREEN,3,playerx+(playerwidth>>1),(playery-playerheight)+COLLISION_BORDER))return 1;
    return 0;
}
